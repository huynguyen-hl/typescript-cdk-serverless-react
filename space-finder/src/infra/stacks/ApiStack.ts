import { Stack, StackProps } from 'aws-cdk-lib';
import { AuthorizationType, CognitoUserPoolsAuthorizer, Cors, LambdaIntegration, MethodOptions, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { IUserPool } from 'aws-cdk-lib/aws-cognito';
import { Construct } from 'constructs';

interface ApiStackProps extends StackProps {
  lambdaIntegration: LambdaIntegration,
  userPool: IUserPool
};

export class ApiStack extends Stack {

  constructor(scope: Construct, id: string, props: ApiStackProps) {
    super(scope, id);

    const api = new RestApi(this, 'SpacesApi', {
      defaultCorsPreflightOptions: {
        allowOrigins: Cors.ALL_ORIGINS,
        allowMethods: Cors.ALL_METHODS
      }
    });

    const authorizer = new CognitoUserPoolsAuthorizer(this, 'SpacesApiAuthorizer', {
      cognitoUserPools: [props.userPool],
      identitySource: 'method.request.header.Authorization',
    });
    authorizer._attachToApi(api);

    const optionsWithAuth: MethodOptions = {
      authorizationType: AuthorizationType.COGNITO,
      authorizer,
    };

    const spacesReousrce = api.root.addResource('spaces');
    spacesReousrce.addMethod('GET', props.lambdaIntegration, optionsWithAuth);
    spacesReousrce.addMethod('POST', props.lambdaIntegration, optionsWithAuth);
    spacesReousrce.addMethod('PUT', props.lambdaIntegration, optionsWithAuth);
    spacesReousrce.addMethod('DELETE', props.lambdaIntegration, optionsWithAuth);

  }
}