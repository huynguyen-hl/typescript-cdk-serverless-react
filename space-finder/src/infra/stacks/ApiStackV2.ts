import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { CorsHttpMethod, HttpApi, HttpMethod } from 'aws-cdk-lib/aws-apigatewayv2';
import { HttpUserPoolAuthorizer } from 'aws-cdk-lib/aws-apigatewayv2-authorizers';
import { HttpLambdaIntegration } from 'aws-cdk-lib/aws-apigatewayv2-integrations';
import { IUserPool } from 'aws-cdk-lib/aws-cognito';
import { Construct } from 'constructs';

interface ApiStackV2Props extends StackProps {
  lambdaIntegration: HttpLambdaIntegration,
  userPool: IUserPool
};

export class ApiStackV2 extends Stack {
  constructor(scope: Construct, id: string, props: ApiStackV2Props) {
    super(scope, id);

    const api = new HttpApi(this, 'SpacesApi', {
      apiName: 'SpacesApi',
      corsPreflight: {
        allowOrigins: ['*'],
        allowMethods: [CorsHttpMethod.ANY]
      },
    });

    const authorizer = new HttpUserPoolAuthorizer('BooksAuthorizer', props.userPool);

    api.addRoutes({
      path: '/spaces',
      methods: [HttpMethod.GET, HttpMethod.POST, HttpMethod.PUT, HttpMethod.DELETE],
      integration: props.lambdaIntegration,
      authorizer
    });

    new CfnOutput(this, 'SpaceApiUrl', {
      value: api.url
    })
  }
}