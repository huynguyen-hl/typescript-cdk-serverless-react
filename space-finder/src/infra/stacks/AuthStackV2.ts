import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import {
  CfnIdentityPool,
  CfnIdentityPoolRoleAttachment,
  UserPool,
  UserPoolClient,
} from 'aws-cdk-lib/aws-cognito';
import {
  IdentityPool,
  UserPoolAuthenticationProvider,
  IdentityPoolRoleMapping,
  IdentityPoolProviderUrl,
} from 'aws-cdk-lib/aws-cognito-identitypool';
import { Effect, FederatedPrincipal, PolicyStatement, Role } from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

export class AuthStack extends Stack {
  public userPool: UserPool;
  private userPoolClient: UserPoolClient;
  // private userIdentityPool: IdentityPool;
  private userIdentityPool: CfnIdentityPool;
  private authenticatedRole: Role;
  private unauthenticatedRole: Role;
  private adminRole: Role;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    this.createUserPool();
    this.createUserPoolClient();
    this.createUserIdentityPool();
    this.createRoles();
    this.createUserPoolAdminGroup();
    this.attachRolesToIdentityPool();
    // this.addPoliciesToIdentityRoles();
  }

  private createUserPool() {
    this.userPool = new UserPool(this, 'SpaceUserPool', {
      userPoolName: 'SpaceUserPool',
      selfSignUpEnabled: true,
      signInAliases: {
        username: true,
        email: true,
      },
    });
    new CfnOutput(this, 'SpaceUserPoolId', {
      value: this.userPool.userPoolId,
    });
  }

  private createUserPoolClient() {
    this.userPoolClient = this.userPool.addClient('SpaceUserPoolClient', {
      authFlows: {
        adminUserPassword: true,
        custom: true,
        userPassword: true,
        userSrp: true,
      },
    });

    new CfnOutput(this, 'SpaceUserPoolClientId', {
      value: this.userPoolClient.userPoolClientId,
    });
  }

  private createUserPoolAdminGroup() {
    this.userPool.addGroup('AdminGroup', {
      groupName: 'admins',
      description: 'Group for admin users',
      role: this.adminRole
    });
  }

  private createUserIdentityPool() {
    this.userIdentityPool = new CfnIdentityPool(this, 'SpaceUserIdentityPool', {
      allowUnauthenticatedIdentities: true,
      cognitoIdentityProviders: [
        {
          clientId: this.userPoolClient.userPoolClientId,
          providerName: this.userPool.userPoolProviderName,
        },
      ],
    });
    new CfnOutput(this, 'SpaceUserIdentityPoolId', {
      value: this.userIdentityPool.ref,
    });
    // this.userIdentityPool = new IdentityPool(this, 'SpaceUserIdentityPool', {
    //   identityPoolName: 'SpaceUserIdentityPool',
    //   allowUnauthenticatedIdentities: true,
    // });

    // this.userIdentityPool.addUserPoolAuthentication(
    //   new UserPoolAuthenticationProvider({
    //     userPool: this.userPool,
    //     userPoolClient: this.userPoolClient,
    //   })
    // );

    // new CfnOutput(this, 'SpaceUserIdentityPoolId', {
    //   value: this.userIdentityPool.identityPoolId,
    // });
  }

  private createRoles() {
    this.authenticatedRole = new Role(this, 'CognitoDefaultAuthenticatedRole', {
      assumedBy: new FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: {
            'cognito-identity.amazonaws.com:aud': this.userIdentityPool.ref,
          },
          'ForAnyValue:StringLike': {
            'cognito-identity.amazonaws.com:amr': 'authenticated',
          },
        },
        'sts:AssumeRoleWithWebIdentity'
      ),
    });

    this.unauthenticatedRole = new Role(this, 'CognitoDefaultUnauthenticatedRole', {
      assumedBy: new FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: {
            'cognito-identity.amazonaws.com:aud': this.userIdentityPool.ref,
          },
          'ForAnyValue:StringLike': {
            'cognito-identity.amazonaws.com:amr': 'unauthenticated',
          },
        },
        'sts:AssumeRoleWithWebIdentity'
      ),
    });

    this.adminRole = new Role(this, 'CognitoAdminRole', {
      assumedBy: new FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: {
            'cognito-identity.amazonaws.com:aud': this.userIdentityPool.ref,
          },
          'ForAnyValue:StringLike': {
            'cognito-identity.amazonaws.com:amr': 'authenticated',
          },
        },
        'sts:AssumeRoleWithWebIdentity'
      ),
    });
    this.adminRole.addToPolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      actions: [
        's3:Get*',
        's3:ListAllMyBuckets',
        's3:Describe*',
      ],
      resources: ['*']
    }));
  }

  private attachRolesToIdentityPool() {
    new CfnIdentityPoolRoleAttachment(this, 'SpaceUserIdentityPoolRoleAttachment', {
      identityPoolId: this.userIdentityPool.ref,
      roles: {
        authenticated: this.authenticatedRole.roleArn,
        unauthenticated: this.unauthenticatedRole.roleArn,
      },
      roleMappings: {
        adminsMapping: {
          type: 'Token',
          ambiguousRoleResolution: 'AuthenticatedRole',
          identityProvider: `${this.userPool.userPoolProviderName}:${this.userPoolClient.userPoolClientId}`,
        },
      },
    });
  }
}
