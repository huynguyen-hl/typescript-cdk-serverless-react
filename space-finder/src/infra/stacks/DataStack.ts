import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { AttributeType, ITable, Table } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';
import { getSuffixFromStack } from '../Utils';
import { Bucket, HttpMethods, ObjectOwnership } from 'aws-cdk-lib/aws-s3';

export class DataStack extends Stack {
  public readonly spacesTable: ITable;
  public readonly photosBucket: Bucket;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id);

    const suffix = getSuffixFromStack(this);

    this.photosBucket = new Bucket(this, 'SpaceFinderPhotos', {
      bucketName: `space-finder-photos-${suffix}`,
      objectOwnership: ObjectOwnership.OBJECT_WRITER,
      cors: [{
        allowedMethods: [
          HttpMethods.HEAD,
          HttpMethods.GET,
          HttpMethods.PUT,
        ],
        allowedOrigins: ['*'],
        allowedHeaders: ['*']
      }],
      blockPublicAccess: {
        blockPublicAcls: false,
        blockPublicPolicy: false,
        ignorePublicAcls: false,
        restrictPublicBuckets: false,
      }
    });
    new CfnOutput(this, 'SpaceFinderPhotosBucketName', {
      value: this.photosBucket.bucketName
    });

    this.spacesTable = new Table(this, 'SpacesTable', {
      partitionKey: {
        name: 'id',
        type: AttributeType.STRING,
      },
      readCapacity: 1,
      writeCapacity: 1,
      tableName: `SpacesTable-${suffix}`
    });
  }
}