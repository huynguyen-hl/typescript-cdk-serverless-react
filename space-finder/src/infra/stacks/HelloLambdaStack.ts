import { fileURLToPath } from 'url';
import { join, dirname } from 'path';
import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Code, Runtime } from 'aws-cdk-lib/aws-lambda';
import { LambdaIntegration } from 'aws-cdk-lib/aws-apigateway';
import { ITable } from 'aws-cdk-lib/aws-dynamodb';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

interface LambdaStackProps extends StackProps {
  spacesTable: ITable;
}

export class HelloLambdaStack extends Stack {
  public readonly helloLambdaIntegration: LambdaIntegration;

  constructor(scope: Construct, id: string, props: LambdaStackProps) {
    super(scope, id);

    const helloLambda = new NodejsFunction(this, 'HelloLambda', {
      runtime: Runtime.NODEJS_22_X,
      handler: 'handler',
      entry: join(__dirname, '..', '..', 'services', 'hello.ts'),
      environment: {
        SPACES_TABLE: props.spacesTable.tableName,
      },
    });

    helloLambda.addToRolePolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      actions: [
        's3:ListAllMyBuckets',
        's3:ListBucket',
      ],
      resources: ['*'],
    }))

    this.helloLambdaIntegration = new LambdaIntegration(helloLambda);
  }
}