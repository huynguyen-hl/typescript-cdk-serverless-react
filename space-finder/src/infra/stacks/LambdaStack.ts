import { fileURLToPath } from 'url';
import { join, dirname } from 'path';
import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Code, Runtime, Tracing } from 'aws-cdk-lib/aws-lambda';
import { LambdaIntegration } from 'aws-cdk-lib/aws-apigateway';
import { ITable } from 'aws-cdk-lib/aws-dynamodb';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { HttpLambdaIntegration } from 'aws-cdk-lib/aws-apigatewayv2-integrations';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

interface LambdaStackProps extends StackProps {
  spacesTable: ITable;
}

export class LambdaStack extends Stack {
  public readonly lambdaIntegration: LambdaIntegration;
  // public readonly lambdaIntegration: HttpLambdaIntegration;

  constructor(scope: Construct, id: string, props: LambdaStackProps) {
    super(scope, id);

    const spacesLambda = new NodejsFunction(this, 'SpacesLambda', {
      runtime: Runtime.NODEJS_22_X,
      handler: 'handler',
      entry: join(__dirname, '..', '..', 'services', 'spaces', 'handler.ts'),
      environment: {
        SPACES_TABLE: props.spacesTable.tableName,
      },
      tracing: Tracing.ACTIVE,
      timeout: Duration.minutes(1)
    });

    spacesLambda.addToRolePolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      actions: [
        'dynamodb:PutItem',
        'dynamodb:GetItem',
        'dynamodb:DeleteItem',
        'dynamodb:UpdateItem',
        'dynamodb:Scan',
      ],
      resources: [props.spacesTable.tableArn],
    }));

    this.lambdaIntegration = new LambdaIntegration(spacesLambda);
    // this.lambdaIntegration = new HttpLambdaIntegration('SpacesLambdaIntegration', spacesLambda);
  }
}