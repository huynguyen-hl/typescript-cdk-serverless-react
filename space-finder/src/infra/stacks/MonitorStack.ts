import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Alarm, Metric, Unit } from 'aws-cdk-lib/aws-cloudwatch';
import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { EmailSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';
import { Construct } from 'constructs';

export class MonitorStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const spaceApi4xxAlarm = new Alarm(this, 'spaceApi4xxAlarm', {
      alarmName: 'spaceApi4xxAlarm',
      metric: new Metric({
        metricName: '4XXError',
        namespace: 'AWS/ApiGateway',
        period: Duration.minutes(1),
        statistic: 'Sum',
        unit: Unit.COUNT,
        dimensionsMap: {
          ApiName: 'SpacesApi',
        },
      }),
      evaluationPeriods: 1,
      threshold: 5,
    });

    const spaceApiTopic = new Topic(this, 'ApiSpaceSnsTopic', {
      topicName: 'ApiSpaceSnsTopic',
      fifo: false,
    });
    spaceApiTopic.addSubscription(new EmailSubscription('scythedead.h@gmail.com'))

    spaceApi4xxAlarm.addAlarmAction(new SnsAction(spaceApiTopic));

  }
}
