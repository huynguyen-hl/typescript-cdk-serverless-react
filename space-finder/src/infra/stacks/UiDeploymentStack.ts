import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { existsSync } from 'fs';
import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
import { getSuffixFromStack } from '../Utils';
import { BucketDeployment, Source } from 'aws-cdk-lib/aws-s3-deployment';
import { Distribution, OriginAccessIdentity } from 'aws-cdk-lib/aws-cloudfront';
import { S3BucketOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export class UiDeploymentStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const stackSuffix = getSuffixFromStack(this);

    const uiDeploymentBucket = new Bucket(this, 'SpaceFinderFrontendDeploymentBucket', {
      bucketName: `space-finder-frontend-${stackSuffix}`,
    });

    const spaceFinderFrontendDir = join(
      __dirname,
      '..',
      '..',
      '..',
      'space-finder-frontend',
      'dist'
    );
    if (!existsSync(spaceFinderFrontendDir)) {
      console.warn(`Ui dir not found: ${spaceFinderFrontendDir}`);
      return;
    }

    new BucketDeployment(this, 'SpaceFinderDeploymentbucketDeployment', {
      destinationBucket: uiDeploymentBucket,
      sources: [Source.asset(spaceFinderFrontendDir)],
    });

    const originIdentity = new OriginAccessIdentity(this, 'SpaceFinderOAI');

    const spaceFinderUiDistribution = new Distribution(this, 'SpaceFinderDistribution', {
      defaultRootObject: 'index.html',
      defaultBehavior: {
        origin: S3BucketOrigin.withOriginAccessIdentity(uiDeploymentBucket, {
          originAccessIdentity: originIdentity,
        }),
      },
    });

    new CfnOutput(this, 'SpaceFinderCdnUrl', {
      value: spaceFinderUiDistribution.distributionDomainName,
    });
  }
}
