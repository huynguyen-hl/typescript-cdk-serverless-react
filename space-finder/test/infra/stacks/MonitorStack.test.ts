import { App, Resource } from 'aws-cdk-lib';
import { MonitorStack } from '../../../src/infra/stacks/MonitorStack';
import { Capture, Match, Template } from 'aws-cdk-lib/assertions';

describe('MonitorStack test suite', () => {
  let monitorStackTemplate: Template;
  let snsTopicResource: ReturnType<typeof monitorStackTemplate.findResources>;
  let emailSubscriptionResource: ReturnType<typeof monitorStackTemplate.findResources>;
  let cloudwatchAlarmResource: ReturnType<typeof monitorStackTemplate.findResources>;

  beforeAll(() => {
    const app = new App({
      outdir: 'cdk.out',
    });

    const monitorStack = new MonitorStack(app, 'MonitorStack');
    monitorStackTemplate = Template.fromStack(monitorStack);
    snsTopicResource = monitorStackTemplate.findResources('AWS::SNS::Topic');
    emailSubscriptionResource = monitorStackTemplate.findResources('AWS::SNS::Subscription');
    cloudwatchAlarmResource = monitorStackTemplate.findResources('AWS::CloudWatch::Alarm');
  });

  it('should correct SNS topic config', () => {
    monitorStackTemplate.hasResourceProperties('AWS::SNS::Topic', {
      FifoTopic: false,
      TopicName: 'ApiSpaceSnsTopic',
    });
  });

  it('should correct SNS subscription config - with Matcher', () => {
    monitorStackTemplate.hasResourceProperties(
      'AWS::SNS::Subscription',
      Match.objectEquals({
        Endpoint: 'scythedead.h@gmail.com',
        Protocol: 'email',
        TopicArn: {
          Ref: Match.stringLikeRegexp('ApiSpaceSnsTopic'),
        },
      })
    );
  });

  it('should correct CloudWatch Alarm config - with Matcher', () => {
    monitorStackTemplate.hasResourceProperties(
      'AWS::CloudWatch::Alarm',
      Match.objectEquals({
        AlarmActions: [
          {
            Ref: Match.stringLikeRegexp('ApiSpaceSnsTopic'),
          },
        ],
        AlarmName: 'spaceApi4xxAlarm',
        ComparisonOperator: 'GreaterThanOrEqualToThreshold',
        Dimensions: [
          {
            Name: 'ApiName',
            Value: 'SpacesApi',
          },
        ],
        EvaluationPeriods: 1,
        MetricName: '4XXError',
        Namespace: 'AWS/ApiGateway',
        Period: 60,
        Statistic: 'Sum',
        Threshold: 5,
        Unit: 'Count',
      })
    );
  });

  it('should correct SNS subscription config - with exact', () => {
    const snsTopicGeneratedName = Object.keys(snsTopicResource)[0];

    monitorStackTemplate.hasResourceProperties('AWS::SNS::Subscription', {
      Endpoint: 'scythedead.h@gmail.com',
      Protocol: 'email',
      TopicArn: {
        Ref: snsTopicGeneratedName,
      },
    });
  });

  it('should correct CloudWatch Alarm config - with exact', () => {
    const snsTopicGeneratedName = Object.keys(snsTopicResource)[0];

    monitorStackTemplate.hasResourceProperties(
      'AWS::CloudWatch::Alarm',
      Match.objectEquals({
        AlarmActions: [
          {
            Ref: snsTopicGeneratedName,
          },
        ],
        AlarmName: 'spaceApi4xxAlarm',
        ComparisonOperator: 'GreaterThanOrEqualToThreshold',
        Dimensions: [
          {
            Name: 'ApiName',
            Value: 'SpacesApi',
          },
        ],
        EvaluationPeriods: 1,
        MetricName: '4XXError',
        Namespace: 'AWS/ApiGateway',
        Period: 60,
        Statistic: 'Sum',
        Threshold: 5,
        Unit: 'Count',
      })
    );
  });

  it('should correct AlarmActions config', () => {
    const alarmActionsCapture = new Capture();
    monitorStackTemplate.hasResourceProperties('AWS::CloudWatch::Alarm', {
      AlarmActions: alarmActionsCapture,
    });

    expect(alarmActionsCapture.asArray()).toEqual([
      {
        Ref: expect.stringMatching(/^ApiSpaceSnsTopic/),
      },
    ]);
  });

  it('should match MonitorStack snapshot', () => {
    expect(monitorStackTemplate.toJSON()).toMatchSnapshot();
  });

  it('should match SNS Topic snapshot', () => {
    expect(snsTopicResource).toMatchSnapshot();
  });

  it('should match SNS Email Subscription snapshot', () => {
    expect(emailSubscriptionResource).toMatchSnapshot();
  });

  it('should match Cloudwatch Alarm snapshot', () => {
    expect(cloudwatchAlarmResource).toMatchSnapshot();
  });
});
