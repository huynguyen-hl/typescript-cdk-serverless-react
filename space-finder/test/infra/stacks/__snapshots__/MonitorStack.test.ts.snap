// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MonitorStack test suite should match Cloudwatch Alarm snapshot 1`] = `
{
  "spaceApi4xxAlarm8E4FAA1F": {
    "Properties": {
      "AlarmActions": [
        {
          "Ref": "ApiSpaceSnsTopic4F7D337B",
        },
      ],
      "AlarmName": "spaceApi4xxAlarm",
      "ComparisonOperator": "GreaterThanOrEqualToThreshold",
      "Dimensions": [
        {
          "Name": "ApiName",
          "Value": "SpacesApi",
        },
      ],
      "EvaluationPeriods": 1,
      "MetricName": "4XXError",
      "Namespace": "AWS/ApiGateway",
      "Period": 60,
      "Statistic": "Sum",
      "Threshold": 5,
      "Unit": "Count",
    },
    "Type": "AWS::CloudWatch::Alarm",
  },
}
`;

exports[`MonitorStack test suite should match MonitorStack snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApiSpaceSnsTopic4F7D337B": {
      "Properties": {
        "FifoTopic": false,
        "TopicName": "ApiSpaceSnsTopic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "ApiSpaceSnsTopicscythedeadhgmailcom17D5A525": {
      "Properties": {
        "Endpoint": "scythedead.h@gmail.com",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "ApiSpaceSnsTopic4F7D337B",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "spaceApi4xxAlarm8E4FAA1F": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "ApiSpaceSnsTopic4F7D337B",
          },
        ],
        "AlarmName": "spaceApi4xxAlarm",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": "SpacesApi",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "4XXError",
        "Namespace": "AWS/ApiGateway",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 5,
        "Unit": "Count",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`MonitorStack test suite should match SNS Email Subscription snapshot 1`] = `
{
  "ApiSpaceSnsTopicscythedeadhgmailcom17D5A525": {
    "Properties": {
      "Endpoint": "scythedead.h@gmail.com",
      "Protocol": "email",
      "TopicArn": {
        "Ref": "ApiSpaceSnsTopic4F7D337B",
      },
    },
    "Type": "AWS::SNS::Subscription",
  },
}
`;

exports[`MonitorStack test suite should match SNS Topic snapshot 1`] = `
{
  "ApiSpaceSnsTopic4F7D337B": {
    "Properties": {
      "FifoTopic": false,
      "TopicName": "ApiSpaceSnsTopic",
    },
    "Type": "AWS::SNS::Topic",
  },
}
`;
